### DISCUSSION TITLE
Is it possible to have deterministic build output?

### DISCUSSION THREAD

**Discussion by scameron** [webpack/webpack#16899] [General]
Hello,

Every time I run our build, if I look at the size of the output chunk files, it is different every single time.  Usually the number of chunk files is the same, but the size on disk of the files is different.  None of the source code or settings have changed in between builds.  This is just running the same build multiple times.

Here are the total chunk sizes from 8 runs of our build, one after another, in bytes:

- 200011996
- 200027705
- 200011666
- 200020535
- 200016600
- 200020336
- 200050837

The difference between the biggest and smallest is 39171 bytes.  Percentage-wise this is totally insignificant, like 0.02% of the total build size.  So it's not a problem for the build itself.

The problem is that we are trying to implement build checks that detect when somebody has made a change that is causing the build output to increase.  And with non-deterministic output, it makes it difficult to pin down a specific change to blame, so we can catch it before it gets merged.  A commit that increases the size, might make it under the current threshold, and then that code could accidentally fail the check in somebody else's PR because they happen to be unlucky and hit a build that comes in on the high side.

We could implement our check based on module sizes, but that doesn't take into consideration the duplication that can happen during chunk splitting.  We're hoping to have out check be done on the true final output size of the build.

Any ideas about why this non-deterministic output might be happening?   I'm guessing it's during splitChunks, but I don't know for sure.

**Comment by alexander-akait** [comment]
Can you provide your configuration?

**Reply by scameron** [reply]
It's a bit complex, is there any part in particular you're interested in?  Or the whole thing?

**Reply by alexander-akait** [reply]
Plugins, because I think the problem is there

**Reply by scameron** [reply]
After your reply, I thought I should really try to narrow down the source of the problem.  In the end, it looks like it's due to a single `optimization` setting: `moduleIds: "size"`.  With that setting, each time I run the build it results in a slightly different output.

It looks like "size" tries to set module Id in order of module occurrence in the build.  Is it expected that this order is not stable when the code has not been touched?

**Reply by alexander-akait** [reply]
Do you use different platforms - windows/linux/macos or only single? What about `\r\n` and `\n` git configuration?

**Reply by scameron** [reply]
I'm just running `npm run build --production` multiple times on the same Windows machine without touching the code.

**Reply by alexander-akait** [reply]
Oh, without reproducible repo it really hard to say, but yes when you have `moduleIds: "size"` output can be non determenistic

**Reply by scameron** [reply]
It tried `moduleIds: "deterministic"` and of course that's stable, but it increases our total build size by nearly 500KB.  And we don't really need that much stability (across code changes) anyway.

It seems like "size" should be able to use a stable sort to produce the same results every time as long as the code doesn't change.  Maybe I'll try to debug a bit further tomorrow.

**Reply by alexander-akait** [reply]
Yeah, feel free to ping me

**Reply by scameron** [reply]
I found the problem, but I don't understand the root cause.

In `OccurrenceModuleIdsPlugin` it applies a [sort to modulesInOccurrenceOrder](https://github.com/webpack/webpack/blob/main/lib/ids/OccurrenceModuleIdsPlugin.js#L135) but in many cases the `occursInInitialChunksMap` lookup returns `NaN` and since all comparisons with `NaN` are falsy, the sort becomes non-deterministic.

How does `occursInInitialChunksMap` end up with `NaN` values in it?  The value comes, in part, from [countOccursInEntry](https://github.com/webpack/webpack/blob/main/lib/ids/OccurrenceModuleIdsPlugin.js#L76) where sometimes this line returns `undefined`:

    initialChunkChunkMap.get(originModule);

The reason is that `originModule`, returned from `moduleGraph.getIncomingConnectionsByOriginModule` does not exist in the `initialChunkChunkMap`.  Looking a bit further back, it also doesn't exist in the `modulesInOccurrenceOrder` array.

The `modulesInOccurrenceOrder` array is populated in `IdHelpers.getUsedModuleIdsAndModules` and in there you can see that it only includes modules that are in [at least one chunk](https://github.com/webpack/webpack/blob/main/lib/ids/IdHelpers.js#L261).  My `NaN` modules, for some reason, are in zero chunks.

How can a module be in zero chunks when it is the origin module for incoming connections to other modules (which are in chunks)?

**Reply by alexander-akait** [reply]
Sounds like a bug, I think we missed extra check somewhere, maybe you can isolate it in reproducible example (for future tests) and we will investigate deeply

**Reply by scameron** [reply]
The problem is I'm a little bit shooting in the dark about what is causing these particular modules to belong to zero chunks, but not others?  Any idea what might be special about these?

I did notice that the zero chunk module seems to contain only imports and an export and not much else.  Does that ring any bells?  Maybe around inlining or some other kind of internal optimization?

**Reply by scameron** [reply]
Actually, just thinking a bit more.  You might already have a repro case in your tests and not even know it because there is really no functional breakage here.  Everything functions perfectly fine, you just end up with non-deterministic module Id assignment.

Could you try to run your automated tests and detect if you hit any cases where the `occursInInitialChunksMap` lookup returns `NaN`?
https://github.com/webpack/webpack/blob/main/lib/ids/OccurrenceModuleIdsPlugin.js#L137

Who knows, maybe it's already happening in the existing tests.  In the meantime, I'll keep experimenting to find a simple repro case on my side.

**Reply by alexander-akait** [reply]
Yeah, try to disable optimization at all and enably step by step, so we can undestand what do it

**Reply by scameron** [reply]
At this point, the `optimization` config is completely empty except for `moduleIds: "size"`.  I was just able to also get the `NaN` values in a another (simpler) build.  Would be good if you could try your existing tests to see if it happens there too.

**Reply by alexander-akait** [reply]
I mean set all options inside `optimization` to false (except `moduleIds`), i.e. disable them. Also can you check this in development mode?

**Reply by scameron** [reply]
I tried out your suggestion and it looks like the problem goes away once I set `usedExports` to false.  If `usedExports` is true or "global", I encounter the `NaN` values.

Production vs development doesn't make any difference.

What does it consider to count as "used"?  The `NaN` module is imported by another module and then that module, in turn, is dynamically imported by another module.  Shouldn't that be considered to be "used"?

**Reply by alexander-akait** [reply]
hm, looks like you have `export something from "./test.js"`, right? Seems we should ignore unused modules where, so you will not have `NaN`... Can you try it locally, just extra check in `OccurrenceModuleIdsPlugin`?

**Reply by scameron** [reply]
In the example I'm looking at, I don't see any `export something from "./test.js"`, but it's possible I'm just not looking far enough up the graph.  The `NaN` module itself exports only a single JS object.  That module is imported by another module that exports a single function (returning an object).  And then there is a dynamic import in a 3rd module that imports the whole 2nd module.  Then there are many callers of that dynamic import.

Could you elaborate a bit about what you'd like me to try inside `OccurrenceModeulIdsPlugin`?  It looks like your suggestion might have been truncated...

**Reply by alexander-akait** [reply]
> Could you elaborate a bit about what you'd like me to try inside OccurrenceModeulIdsPlugin?

Just ignore zero chunk modules

**Reply by scameron** [reply]
I made a small modification to `countOccursInEntry` in `OccurrenceModuleIdsPlugin` that seems to work. The NaN values are gone and the build output is completely stable down to the byte now:

    const countOccursInEntry = module => {
        let sum = 0;
        for (const [
             originModule,
             connections
        ] of moduleGraph.getIncomingConnectionsByOriginModule(module)) {
            if (!originModule) continue;
            if (!connections.some(c => c.isTargetActive(undefined))) continue;
            sum += (initialChunkChunkMap.get(originModule) || 0)
        }
        return sum;
    };

The only change I made is to the line `sum += (initialChunkChunkMap.get(originModule) || 0)` where I use a `0` if `originModule` is not found.

**Reply by alexander-akait** [reply]
Yeah, I think it is valid solution, can you send a PR?

**Reply by scameron** [reply]
I created PR https://github.com/webpack/webpack/pull/16933 but it looks like I don't have rights to add reviewers.  Can you add yourself in there?
